# This patch file is from Qualcomm Technologies, Inc. and is provided under the following license:
#
# Copyright (c) 2025 Qualcomm Technologies, Inc. All rights reserved.
#
# SPDX-License-Identifier: BSD-3-Clause-Clear

From f1467b717cd1e2324f4e6f19e9021a0e5832da38 Mon Sep 17 00:00:00 2001
From: Rafal Skierkowski <rskierko@qti.qualcomm.com>
Date: Wed, 21 May 2025 13:54:19 +0200
Subject: [PATCH] Store boot-recovery wipe command for GVM in case of device
 lock state change.

In case device lock state change Android have a problem to boot due to
problem to decrypt userdata partition and reboot to recovery mode with
prompt. Lock state change impact with clearing keymaster keys and change
of root of trust. In such case wipe of data is needed to bring back Android
to bootable state.
Store wipe recovery command in gvm misc partition to inform guest about
necessity of wiping data. Android should boot into recovery mode and
wipe data and reboot itself to normal mode.

Signed-off-by: Ankurkumar Satasiya <asatasiy@qti.qualcomm.com>
---
 .../QcomModulePkg/Library/BootLib/BootLib.inf |  1 +
 .../Library/BootLib/DeviceInfo.c              | 21 +++++++++++++++++++
 .../edk2/QcomModulePkg/QcomModulePkg.dec      |  1 +
 3 files changed, 23 insertions(+)

diff --git a/QcomModulePkg/Library/BootLib/BootLib.inf b/QcomModulePkg/Library/BootLib/BootLib.inf
index 5b855cf..1c58613 100755
--- a/QcomModulePkg/Library/BootLib/BootLib.inf
+++ b/QcomModulePkg/Library/BootLib/BootLib.inf
@@ -116,6 +116,7 @@
 	gEfiLogFSPartitionGuid
 	gEfiNandUserPartitionGuid
 	gEfiMiscPartitionGuid
+	gEfiGvmMiscPartitionGuid
 	gEfiRNGAlgRawGuid
 	gEfiUfsLU0Guid
 	gEfiUfsLU1Guid
diff --git a/QcomModulePkg/Library/BootLib/DeviceInfo.c b/QcomModulePkg/Library/BootLib/DeviceInfo.c
index 1ab7628..278f61f 100644
--- a/QcomModulePkg/Library/BootLib/DeviceInfo.c
+++ b/QcomModulePkg/Library/BootLib/DeviceInfo.c
@@ -164,6 +164,8 @@ SetDeviceUnlockValue (UINT32 Type, BOOLEAN State)
   EFI_STATUS Status = EFI_SUCCESS;
   struct RecoveryMessage Msg;
   EFI_GUID Ptype = gEfiMiscPartitionGuid;
+  EFI_GUID GvmPtype = gEfiGvmMiscPartitionGuid;
+
   MemCardType CardType = UNKNOWN;
 
   switch (Type) {
@@ -204,6 +206,25 @@ SetDeviceUnlockValue (UINT32 Type, BOOLEAN State)
     }
 
     Status = WriteToPartition (&Ptype, &Msg, sizeof (Msg));
+    if (Status != EFI_SUCCESS) {
+      DEBUG ((EFI_D_ERROR, "Fail to write recovery msg: %r", Status));
+      return Status;
+    }
+
+    /* Add boot-recovery command to gvm message to boot into recovery directly */
+    Status = AsciiStrnCpyS (Msg.command, sizeof (Msg.command),
+                          RECOVERY_BOOT_RECOVERY, AsciiStrLen (RECOVERY_BOOT_RECOVERY));
+
+    if (Status != EFI_SUCCESS) {
+      DEBUG ((EFI_D_ERROR, "Fail to prepare command: %r", Status));
+      return Status;
+    }
+
+    Status = WriteToPartition (&GvmPtype, &Msg, sizeof (Msg));
+    if (Status != EFI_SUCCESS) {
+      DEBUG ((EFI_D_ERROR, "Fail to write gvm recovery msg: %r", Status));
+      return Status;
+    }
   }
 
   return Status;
diff --git a/QcomModulePkg/QcomModulePkg.dec b/QcomModulePkg/QcomModulePkg.dec
index 2290b35..8515fd2 100755
--- a/QcomModulePkg/QcomModulePkg.dec
+++ b/QcomModulePkg/QcomModulePkg.dec
@@ -78,6 +78,7 @@
   gEfiLogFSPartitionGuid              = { 0xBC0330EB, 0x3410, 0x4951, { 0xA6, 0x17, 0x03, 0x89, 0x8D, 0xBE, 0x33, 0x72 } }
   gEfiAblFvNameGuid=                    { 0x046fae99, 0xcf2e, 0x49ed, { 0xA6, 0xA8, 0xA1, 0x48, 0x8b, 0x7E, 0x80, 0xd3 } }
   gEfiMiscPartitionGuid =               { 0x04600328, 0x0CC3, 0x47B8, { 0x8E, 0xF1, 0xB1, 0x92, 0xC3, 0x9E, 0x3E, 0x1F } }
+  gEfiGvmMiscPartitionGuid =            { 0xef32a33b, 0xa409, 0x486c, { 0x91, 0x41, 0x9f, 0xfb, 0x71, 0x1f, 0x62, 0x66 } }
   gEfiBootImgPartitionGuid =            { 0x20117f86, 0xe985, 0x4357, { 0xb9, 0xee, 0x37, 0x4b, 0xc1, 0xd8, 0x48, 0x7d } }
   gEfiRecoveryImgPartitionGuid =        { 0x9D72D4E4, 0x9958, 0x42DA, { 0xAC, 0x26, 0xBE, 0xA7, 0xA9, 0x0B, 0x04, 0x34 } }
   gEfiSystemPartitionGuid =             { 0x97D7B011, 0x54DA, 0x4835, { 0xB3, 0xC4, 0x91, 0x7A, 0xD6, 0xE7, 0x3D, 0x74 } }
-- 
2.43.2

